rustls/src/client/client_conn.rs

```diff
@@ -340,6 +362,36 @@ pub(super) mod danger {
         pub fn set_certificate_verifier(&mut self, verifier: Arc<dyn ServerCertVerifier>) {
             self.cfg.verifier = verifier;
         }
+
+        /// Create a new dangerous verifier that skips all certificate verification.
+        /// This verifier would only encrypt connection to prevent eavesdropping,
+        /// but will still allow a man-in-the-middle attacks and connections to unverified
+        /// counterparties - i.e. you trust the network not to be compromised.
+        pub fn new_allow_all_verifier() -> Arc<dyn ServerCertVerifier> {
+            Arc::new(NoCertificateVerification {})
+        }
+
+        /// Create a new dangerous verifier that skips certificate verification.
+        /// Will verify the certificate is valid in the following ways:
+        /// - Signed by a valid root
+        /// - Not Expired
+        ///
+        /// TODO: describe dangers of this verification method, possibly adapting
+        /// from the PostgreSQL SSL info page <https://www.postgresql.org/docs/15/libpq-ssl.html>
+        /// > The difference between verify-ca and verify-full depends on the policy of the root CA.
+        /// > If a public CA is used, verify-ca allows connections to a server that somebody else
+        /// > may have registered with the CA.
+        ///
+        /// In PG, verify-full implies hostname is matched against SAN, or CN if no SAN of type dNSName is present,
+        /// but verify-full is less commonly used, so this verifier would skip this verification.
+        ///
+        /// TODO: The above description is only for the PR discussion, and should be fixed before merging.
+        ///
+        pub fn new_ignore_hostname_verifier() -> Arc<dyn ServerCertVerifier> {
```

Comment:
I guess I'm still not exactly clear on the use case here. Given that rustls on main has support for certificates with IP addresses instead of a host name, it should always be possible to cheapyly/simply verify the identity of the server somehow?

Reply:
Here's my use-case back when I used rustls (I don't anymore, for unrelated reasons): We have a software that has a client/server architecture and is often installed on-premise, inside a user's internal network.

To secure the client/server comms, we use TLS, but we don't use the public PKI because some users do not want to have "publicly known" domain names on their internal network. Instead, we use our own root certificate, and give our clients a leaf certificate that chains back to it.

We don't actually *care* about the hostname. We authenticate the server by having a sort of userId field in the leaf cert. In our model, how we connect doesn't matter - a cert that chains to our root with the right userId is the guarantee that we're talking to the right server. Internal networks don't always have domain names, and sometimes the IP ranges will change because of an internal reorg or something. Having our users re-issue TLS certificates when that happens isn't desirable for us.

I guess "technically", we could use the hostname verification and pass the userId as a hostname. That feels like a terrible hack though, and isn't compatible with how we generate our certs today. Plus, it's not exactly compatible with a lot of libraries that use rusTLS under the hood and don't allow setting a different hostname.

Reply:
Fair enough, thanks for the additional context!

Reply:
So for these use cases I think you should be able to copy the `ServerCertVerifier` impl for `WebPkiVerifier` and leave off the calls to `cert.verify_is_valid_for_subject_name()`. Might be good to factor out the remaining code into a separate method.

Reply:
would it make sense to add a non-accessible bool flag inside the `WebPkiVerifier`, which will only be modified by the `danger().without_hostname_verification()` method?  Or you think there should be two totally separate structs?

Reply:
Yeah, a private flag for this seems reasonable to me although maybe it should be a custom email to make it extra obvious what's going on (along the lines of `enum VerifySubjectName { Always, InsecureIgnore }` or something).

Reply:
@djc it seems if I create the `.danged().without_hostname_verification()`, I will not be able to tell what verifier is being used before that call - because by that time the `ConfigBuilder` instance is gone, and the verifier itself is unknown (I only have a dyn trait of unknown type).  What are your thoughts on how I can migrate the ClientConfig state at that point  to the new `NoHostnameVerification` object?